examples/generative/wgan_gp.py

```diff
@@ -0,0 +1,430 @@
+                d_loss = d_cost + gp * self.gp_weight
+
+            # Get the gradients w.r.t the discriminator loss
+            d_gradient = tape.gradient(d_loss, self.discriminator.trainable_variables)
+            # Update the weights of the discriminator using the discriminator optimizer
+            self.d_optimizer.apply_gradients(
+                zip(d_gradient, self.discriminator.trainable_variables)
+            )
+
+        # Train the generator now.
+        # Get the latent vector
+        random_latent_vectors = tf.random.normal(shape=(batch_size, self.latent_dim))
+        with tf.GradientTape() as tape:
+            # Generate fake images using the generator
+            generated_images = self.generator(random_latent_vectors, training=True)
+            # Get the discriminator logits for fake images
+            gen_img_logits = self.discriminator(generated_images, training=True)
+            # Calculate the generator loss
+            g_loss = self.g_loss_fn(gen_img_logits)
+
+        # Get the gradients w.r.t the generator loss
+        gen_gradient = tape.gradient(g_loss, self.generator.trainable_variables)
+        # Update the weights of the generator using the generator optimizer
+        self.g_optimizer.apply_gradients(
+            zip(gen_gradient, self.generator.trainable_variables)
+        )
+        return {"d_loss": d_loss, "g_loss": g_loss}
+
+
+"""
+## Create a callback that periodically saves generated images
+"""
+
+
+class GANMonitor(keras.callbacks.Callback):
+    def __init__(self, num_img=6, latent_dim=128):
+        self.num_img = num_img
+        self.latent_dim = latent_dim
+
+    def on_epoch_end(self, epoch, logs=None):
+        random_latent_vectors = tf.random.normal(shape=(self.num_img, self.latent_dim))
+        generated_images = self.model.generator(random_latent_vectors)
+        generated_images = (generated_images * 127.5) + 127.5
+
+        for i in range(self.num_img):
+            img = generated_images[i].numpy()
+            img = keras.preprocessing.image.array_to_img(img)
+            img.save("generated_img_{i}_{epoch}.png".format(i=i, epoch=epoch))
+
+
+"""
+## Train the end-to-end model
+"""
+
+# Optimizer for both the networks
+# lr=0.0002, beta_1=0.5 are recommened
+generator_optimizer = keras.optimizers.Adam(lr=0.0002, beta_1=0.5, beta_2=0.9)
+discriminator_optimizer = keras.optimizers.Adam(lr=0.0002, beta_1=0.5, beta_2=0.9)
+
+# Define the loss functions to be used for discrimiator
+# This should be -(real_loss - fake_loss)
+# We will add the gradient penalty later to this loss function
+def discriminator_loss(real_img, fake_img):
+    real_loss = tf.reduce_mean(real_img)
+    fake_loss = tf.reduce_mean(fake_img)
+    return fake_loss - real_loss
+
+
+# Define the loss functions to be used for generator
+def generator_loss(fake_img):
+    return -tf.reduce_mean(fake_img)
+
+
+# epochs to train
+nb_epochs = 20
```

Comment:
The prefix to indicate a number is `num_` (though in this case you could just use `epochs` to be consistent with the `fit` signature).